flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,340 @@
+     */
+  def printTree(): Unit = {
+    printTreeRecur(root)
+  }
+
+  def printTreeRecur(node: Node){
+    if(node.children != null) {
+      for (c <- node.children){
+        printTreeRecur(c)
+      }
+    }else{
+      println("printing tree: n.nodeElements " + node.nodeElements)
+    }
+  }
+
+  /**
+   * Recursively adds an object to the tree
+   * @param queryPoint
+   */
+  def insert(queryPoint: Vector){
+    insertRecur(queryPoint,root)
+  }
+
+  private def insertRecur(queryPoint: Vector,node: Node) {
+    if (node.children == null) {
+      if (node.nodeElements.length < maxPerBox ) {
+        node.nodeElements += queryPoint
+      } else{
+        node.makeChildren()
+        for (o <- node.nodeElements){
+          insertRecur(o, node.children(node.whichChild(o)))
+        }
+        node.nodeElements.clear()
+        insertRecur(queryPoint, node.children(node.whichChild(queryPoint)))
+      }
+    } else{
+      insertRecur(queryPoint, node.children(node.whichChild(queryPoint)))
+    }
+  }
+
+  /**
+   * Used to zoom in on a region near a test point for a fast KNN query.
+   * This capability is used in the KNN query to find k "near" neighbors n_1,...,n_k, from
+   * which one computes the max distance D_s to queryPoint.  D_s is then used during the
+   * kNN query to find all points within a radius D_s of queryPoint using searchNeighbors.
+   * To find the "near" neighbors, a min-heap is defined on the leaf nodes of the leaf
+   * nodes of the minimal bounding box of the queryPoint. The priority of a leaf node
+   * is an appropriate notion of the distance between the test point and the node,
+   * which is defined by minDist(queryPoint),
+   *
+   * @param queryPoint
+   * @return
+   */
+  def searchNeighborsSiblingQueue(queryPoint: Vector): ListBuffer[Vector] = {
+    var ret = new ListBuffer[Vector]
+    // edge case when the main box has not been partitioned at all
+    if (root.children == null) {
+      root.nodeElements.clone()
+    } else {
+      val nodeQueue = new PriorityQueue[(Double, Node)]()(Ordering.by(x => x._1))
+      searchRecurSiblingQueue(queryPoint, root, nodeQueue)
+
+      var count = 0
+      while (count < maxPerBox) {
+        val dq = nodeQueue.dequeue()
+        if (dq._2.nodeElements.nonEmpty) {
+          ret ++= dq._2.nodeElements
+          count += dq._2.nodeElements.length
+        }
+      }
+      ret
+    }
+  }
+
+  /**
+   *
+   * @param queryPoint
+   * @param node
+   * @param nodeQueue defined in searchSiblingQueue, this stores nodes based on their
+   *                  distance to node as defined by minDist
+   */
+  private def searchRecurSiblingQueue(queryPoint: Vector, node: Node,
+                                      nodeQueue: PriorityQueue[(Double, Node)]) {
+    if (node.children != null) {
+      for (child <- node.children; if child.contains(queryPoint)) {
+        if (child.children == null) {
+          for (c <- node.children) {
```

Comment:
Why do you add all `children` if only one `child` contains the `queryPoint`? Do you want to store the neighbouring children in `queryPoint` as well?